### PR TITLE
AKCORE-1: Implemented share group heartbeat (4/N)

### PR DESCRIPTION
**Pull request by apoorvmittal10**
- Share group heartbeat implementation with group epoch records on `__consumer_offsets` topic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 1]
```diff
@@ -1362,8 +1365,141 @@ private CoordinatorResult<ShareGroupHeartbeatResponseData, Record> shareGroupHea
...
+        // The assignment is only provided in the following cases:
+        // 1. The member just joined or rejoined to group (epoch equals to zero);
+        // 2. The member's assignment has been updated.
+        if (memberEpoch == 0 || assignmentUpdated) {
+            response.setAssignment(createShareGroupResponseAssignment(updatedMember));
+        }
+
+        // For a consumer group, this would occur as a result of applying (replaying) a member assignment
+        // record. Because a share group doesn't use assignment records, hence we need to do this here.
+        // TODO: we need to verify if bumping the group epoch is done post updating the member's assignment
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1496]: Yes, we need to resolve this during this sprint.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 2]
```diff
@@ -1431,8 +1567,13 @@ private CoordinatorResult<ShareGroupHeartbeatResponseData, Record> shareGroupLea
             String memberId,
             int memberEpoch
     ) throws ApiException {
-        // TODO: Implement this method.
-        throw new UnsupportedOperationException("Share group leave is not yet implemented.");
+        ShareGroup group = getOrMaybeCreateShareGroup(groupId, false);
+        ShareGroupMember member = group.getOrMaybeCreateMember(memberId, false);
+        List<Record> records = shareGroupFenceMember(group, member);
+        log.info("[GroupId {}] Member {} left the consumer group.", groupId, memberId);
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1573]: "share group".

**Review by AndrewJSchofield — APPROVED** [review]
lgtm with a few minor comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 2 above)
Comment (apoorvmittal10) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Done.

**Pull request merged**